### PR TITLE
Set Scope.User.Id to the InstallationId by default

### PR DESCRIPTION
**Pull request by jamescrosswell**
Resolves https://github.com/getsentry/sentry-dotnet/issues/3302

### CODE REVIEW

**Review by bitsandfoxes — CHANGES_REQUESTED** [review]
I think this is doing it the wrong way around. Now the client and the session manager each hold their own version of an `installationIdHelper`. And the client tries to fix a missing ID after the fact?

The ID is part of the state and should sit on the hub if it's getting used in multiple places, which, if it's getting resolved in the `GlobalSessionManager`, is already sort of happening. In that case, it should be set on the user where applicable.
The client should not need go around trying to cover gaps in the setup.

**Comment by jamescrosswell** [comment]
> I think this is doing it the wrong way around. Now the client and the session manager each hold their own version of an `installationIdHelper`. And the client tries to fix a missing ID after the fact?
> 
> The ID is part of the state and should sit on the hub if it's getting used in multiple places, which, if it's getting resolved in the `GlobalSessionManager`, is already sort of happening. In that case, it should be set on the user where applicable. The client should not need go around trying to cover gaps in the setup.

That sounds good in principal. In practice, I had to delay setting the User.Id until the client since setting it earlier impacts various other functionality. 

Here, for example, the client runs an enricher:
https://github.com/getsentry/sentry-dotnet/blob/c5414597e2fd8f7c38d88af539adcec838636041/src/Sentry/SentryClient.cs#L147

The enricher does this:
https://github.com/getsentry/sentry-dotnet/blob/5bcb9cac8b54f283a4ce5cc16573bfdf2970cdf9/src/Sentry/Internal/Enricher.cs#L79-L82

Setting the User.Id before this point would break the enricher then (as it changes the result of the call to `eventLike.HasUser()`.

I think utlimately the InstallationId is not state that's held by the hub at all... rather it's state that gets initialised the first time the SDK runs for a particular project on a particular device. It sets a unique ID identifying that installation and stores it on the file system. Logically, it's a piece of context that stays the same for the duration of the application's execution so it's similar to some of the other stuff we new up when initialising the SDK/Options. It's accessed by the Global Session Manager and also by the Sentry Client - but neither of these creates/owns/stores that state. I'm thinking maybe we could move it to be a Lazy property on the SentryOptions then (`SentryOptions.InstallationId`). That seems like a more natural place for it to live.

File: src/Sentry/GlobalSessionManager.cs [hunk 1]
```diff
@@ -40,124 +39,7 @@ internal class GlobalSessionManager : ISessionManager
...
-                _options.LogDebug("Resolved installation ID '{0}'.", id);
-            }
-            else
-            {
-                _options.LogDebug("Failed to resolve installation ID.");
-            }
-
-            return _resolvedInstallationId = id;
-        }
+        _installationIdHelper = new InstallationIdHelper(options, _persistenceDirectoryPath);
```
[the remark this answers is not in the document]
Comment (jamescrosswell) [reply src/Sentry/GlobalSessionManager.cs]: True... there wasn't really much option. It had to come out of the `GlobalSessionManager` in order to be reusable 🤷🏻‍♂️

**Review by bitsandfoxes — APPROVED** [review]
Ahh, I like this a lot better! That makes a lot of sense, putting it on the options! Thanks for that!

File: src/Sentry/SentryClient.cs [hunk 2]
```diff
@@ -156,6 +156,9 @@ public void CaptureTransaction(SentryTransaction transaction, Scope? scope, Sent
             }
         }
 
+        // Set the user id to our fallback, if this hasn't already been set
+        transaction.User.Id ??= _options.InstallationId;
+
```
[the remark this answers is not in the document]
Comment (jamescrosswell) [reply src/Sentry/SentryClient.cs:161]: We could. The Enricher appears to be applied here (for events):
https://github.com/getsentry/sentry-dotnet/blob/777d6649653b3361fd21b1c547316232442474bd/src/Sentry/Internal/MainSentryEventProcessor.cs#L157

And here (for transactions):
https://github.com/getsentry/sentry-dotnet/blob/37eb4a54cee682bec0d63435ce9959a67fb4656e/src/Sentry/SentryClient.cs#L145

This, in itself, is a bit odd. Why use the `Sentry.Client._enricher` to enrich transactions but use the `MainEventProcessor._enricher` to enrich events? It may be something to do with the order in which things get processed. In ASP.NET Core applications, for example, there is [logic in the ScopeExtensions](https://github.com/getsentry/sentry-dotnet/blob/1447b5598a6c5b23bffc7e3bb19deadb98088acc/src/Sentry.AspNetCore/ScopeExtensions.cs#L39-L51) (for Sentry instrumented applications) and in an [AspNetCoreEnricher](https://github.com/getsentry/sentry-dotnet/blob/bd62cf0421cc7a5d7b7804f8461fae00d47d43b8/src/Sentry.OpenTelemetry/AspNetCoreEnricher.cs) (for OpenTelemetry instrumented applications) that will only be applied if none of the properties for the `User` have been set. We would want that logic to be applied before setting the `User.Id` fallback.

In any event, we could move the logic from `SentryClient` to the `Enricher`. A couple of implications of doing this:
- The logic would then be applied before any transaction processors (that's when the enricher is applied for transactions). 
- We can't guarantee whether the logic gets applied before or after any custom event processors (the `MainSentryEventProcessor` that enriches events is just one of the event processors our SDK applies in a loop).

TLDR; I think it would make the code tidier but the behaviour less certain.

Do we still want to move it?
